Add dropzone

abannerjee opened the pull request.

Commit:
Add 'dropzone' npm module and types to project.

Add the dropzone npm package and DefinitelyTyped dropzone types
to s3commander.

Commit:
Add dropzone controller, component and template to the project.

Using the upload form as the basis, create a dropzone component,
controller and template and add to the project.

Commit:
Create a dropzone directive to handle dropzone events.

Initialize the dropzone element in the directive and add event
handlers to handle the 'sending' event which triggers on each
file upload.

Modify the 'sending' event to append the 'key' form data required
for aws POSTs.

Commit:
Add dropzone component and controller to index.ts and to bucket.

Initialize dropzone component and controller in index.ts and add
a new section in the bucket template for the dropzone element.

Commit:
Increment s3commander version to 0.6.0

abannerjee commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add DropZone to s3commander rewrite](https://trello.com/c/UboCJxye/2363-add-dropzone-to-s3commander-rewrite)

CtrlC-Root commented on src/dropzone/dropzone.html:
```diff
@@ -0,0 +1,12 @@
+<form submit-form
+  action="{{ $ctrl.url }}"
+  class="dropzone"
+  id="s3c-dropzone"
```
This should probably be a `class` instead of an `id`. S3Commander is intended ot be embedded in other pages, you shouldn't assume there's only one instance per page.

abannerjee replied:
Ah, that makes sense. I'll drop this id attribute (it's not being used anywhere).

I will also go ahead and remove the uploadForm component on this branch as well.

CtrlC-Root commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -0,0 +1,45 @@
+
+    let eventHandlers = {
+      'sending': (file, xhr, formData) => {
+        // append AWS upload key to the form data
```
It's not a big deal because the reference is in a comment but keep in mind this code should work with all backends.

abannerjee replied:
Yeah, I agree as well. I know currently we only have the S3 backend defined and the 'key' form data is required to make POSTs on S3. I'll update the comment here to make note this is specific for AWS S3 (and if we have additional backends, we may need to pass that info to the dropzone controller as well).

CtrlC-Root commented on src/bucket/bucket.html:
```diff
@@ -45,4 +45,13 @@
+
+  <!-- dropzone -->
+  <div class="s3c-bottom-section"
+    ng-if='$ctrl.uploadConfig.url'>
```
Why is this `ng-if` newcessary here?

abannerjee replied:
Yeah, this `ng-if` is necessary to allow the dropzone element to be initialized after `uploadConfig` is set and `url` is not undefined.

I am using the `uploadConfig.url` in the dropzone config. Before, the dropzone element would initialize while `uploadConfig` was undefined, and the url would not be set properly in the dropzone config (causing the POST to fail).

mliudev replied:
Wouldn't this prevent this div from loading if `$ctrl.uploadConfig.url` was not defined or is there something that reloads this div once it is defined?

CtrlC-Root replied:
I see. I think it would be better if you tested for `uploadConfig` instead of `uploadConfig.url` though since you pass the former to your control and not the latter.

abannerjee replied:
Good point, this has been changed so `uploadConfig` is tested prior to initializing the dropzone element.

CtrlC-Root reviewed (changes requested):
One change requested and one question in code specific comments. I would also go ahead and remove the `uploadForm` component and related files entirely since we do not intend to use it in the future.

Commit:
Bind backend object in the DropzoneDirective.

The backend object created in BucketController is binded to the
Dropzone directive so it may be used to determine which backend
is being used (currently there is only one which is 's3').

This information is needed by the directive so the form data may
be modified to allow successfully making POSTs when uploading
files to S3.

Commit:
Add function to AmazonS3Backend for setting form data.

When dropzone attempts to upload a file to S3, the form data needs
to be modified to include a 'key' parameter.

In this function, the 'key' parameter is set and the filepath takes into
account which folder the file is uploaded to and/or if the uploaded file
is located within a folder/sub-folder.

Commit:
Update dropzone config.

The dropzone config is updated with the following paramters:

- maxFileSize : This is set to 1GB as per the initial s3commander
                dropzone settings.

- addRemoveLinks: Set to 'True' to allow a link to 'Cancel' a file
                  upload.

- dictCancelUpdate: Names the link to cancel uploads to 'Cancel'.

- dictDefaultMessage: Default message updated to inform users they
                      can click on the dropzone area to bring up the
                      file browser to upload files.

Commit:
Refresh the bucket after successfully uploading files.

The 'success' event handler is added which removes files from
dropzone after successfully uploading them as well as refreshing
the s3commander list.

The 'queuecomplete' event handler is added which calls the binded
'onRefresh' function defined in the controller.

Commit:
Remove UploadForm, FileInput and SubmitForm from s3commander.

The UploadForm component and FileInput, SubmitForm directives
have been replaced with Dropzone and are no longer used within
s3commander.

mliudev commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -24,15 +24,40 @@ export class DropzoneDirective implements ng.IDirective {
     element: ng.IAugmentedJQuery,
     attrs: ng.IAttributes
   ) {
+    let uploadToS3 = (ctrl, file, formData) => {
```
Would it make more sense for this function to be in the AmazonS3Backend.ts file instead? That way this directive could just call something like `backend.upload()`.

abannerjee replied:
This function is misleading and has been updated. The actual upload doesn't occur in the function, but rather what is happening here is the form data is updated and a 'key' value is added (needed to make a POST to aws).

Anyways, this has been moved into AmazonS3Backends as you suggested and the function has been reworded to 'updateFormData'.

mliudev commented on src/index.ts:
```diff
@@ -1,12 +1,10 @@
 import * as angular from 'angular';
 
-import {FileInputDirective} from './input/FileInputDirective';
```
This is removed but the FileInputDirective.ts still exists. On that note, it looks like references to SubmitFormDirective should also be removed.

abannerjee replied:
Thanks for catching this, fileinputdirective and submitformdirective have also been removed in that commit.

mliudev reviewed (changes requested):
Left a couple of minor comments and a change request. Nice job on figuring out how to pass scope to directives.

CtrlC-Root commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -24,15 +24,40 @@ export class DropzoneDirective implements ng.IDirective {
-        formData.append('key', file.name);
+        scope.$apply(() => {
+          // s3 file upload
+          if (scope.$ctrl.backend.hasOwnProperty('s3')) {
```
Don't do this. First, all backend-specific code should be *in* the backend class itself. Second, I think conditional logic around class attributes is a bad way to accomplish this, especially since the `s3` attribute is private. What if I change the API later?

abannerjee replied:
I've added a new commit where I move the 'uploadToS3' function found in the dropzone directive into AwsS3Backend.

abannerjee replied:
The new commit has been rebased into this one.

CtrlC-Root commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -62,6 +62,17 @@ export class DropzoneDirective implements ng.IDirective {
+        // todo: bit of a hack, using querySelector to click 'Refresh'
+        // button instead of directly calling loadContents function
+        // defined in BucketController
+        let query = '.s3c-button[ng-click="$ctrl.loadContents()"]';
```
Why don't you pass a method to the directive from the controller it can use to trigger this instead?

abannerjee replied:
I've tried a few things but still unable to access the 'loadContents' function in the scope of the dropzone directive. I've left this work around for now, but if you have any suggestions on where to make those changes, I can certainly add that here.

CtrlC-Root replied:
I'll take a look.

abannerjee replied:
Thanks for sending me a diff of a different approach for fixing this issue (by using expression binding '&' to bind the parent scope's "loadContents" function). This change has been applied to the dropzone directive.

CtrlC-Root reviewed (changes requested):
OVerall it looks good but there's one question and one issue I'd like to see addressed.

cworth-gh commented on src/dropzone/dropzone.html:
```diff
@@ -1,7 +1,6 @@
 <form submit-form
   action="{{ $ctrl.url }}"
   class="dropzone"
-  id="s3c-dropzone"
```
This might as well be "fixup"ped into the commit that originally introduced this code.

There's no reason to have a commit that introduces a problem, and then a separate commit to fix that problem, as part of a single feature branch to merge.

abannerjee replied:
That makes sense to me, this commit has been merged into the commit where dropzone.html is introduced. Thanks for catching this.

cworth-gh commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -57,10 +38,8 @@ export class DropzoneDirective implements ng.IDirective {
-            uploadToS3(scope.$ctrl, file, formData);
-          }
+          // set form data prior to submitting the dz form
+          scope.$ctrl.backend.setFormData(scope.$ctrl.folder, file, formData);
```
This commit also looks like it should be "fixup"ped into the previous commit in this branch where Alex pointed out the original problem.

abannerjee replied:
Thanks for catching this as well. I've rebased this commit into the previous commit where we add the new function in AmazonS3Backend.
